File: packages/core/src/games/groups.ts
import { bankIndexForDay, bankIndexForSeed, bankDayIndex, bankHolidayPick, type HolidayTable } from '../bank';
import { mulberry32, simpleHash } from '../seed';

/**
 * Kindred — groups of four (More Games §14). Sixteen words hide four groups
 * of four; find them all with at most four mistakes. Tiers 1–4 run from a
 * plain category to wordplay and are shown as one to four pips on the solved
 * bar (never color alone). Submitting four words: a group → it locks and its
 * bar appears; three from one group → "One away"; otherwise a plain miss. A
 * set already tried is free to try again (no second mistake). Four mistakes
 * lose the puzzle and the remaining groups are revealed.
 *
 * Hints (founder, 2026-09-21: "the same two capsules as the other games"):
 * NAME A CATEGORY (1 hint, 100) shows the label of the easiest unsolved group;
 * SHOW A PAIR (2 hints, 200) rings two words that belong together. Neither
 * costs a mistake; both are cheaper than a wrong guess (250) so asking beats
 * guessing blind, and an unaided solve outranks a hinted one.
 *
 * Result row: guess_count = submissions on a win (4 perfect … 7 worst) and
 * groups found + 4 on a loss; boards_solved = groups found of 4. Event sigils
 * (§11): "+t:W1,W2,W3,W4" solved tier t, "x1:…" one away, "x0:…" miss,
 * "=…" a repeated set (free), "?ct" category hint, "?p:A,B" pair hint,
 * "~n" shuffle n. Tile order comes from mulberry32(simpleHash(seed + "-groups-v1"))
 * so every platform deals the same board.
 *
 * Parity-critical: Groups.swift / Groups.kt reproduce this exactly;
 * groups-fixtures.json pins all three.
 */

export const GROUPS_DAILY_EPOCH = '2026-09-23';
export const GROUPS_MAX_MISTAKES = 4;
export const GROUPS_TOTAL_BOARDS = 4;
export const GROUPS_PERFECT_GUESSES = 4;

export interface GroupsGroup { tier: number; label: string; words: string[] }
export interface GroupsPuzzle { id: string; groups: GroupsGroup[]; holiday?: string }
export interface GroupsBank { version: number; epoch: string; daily: GroupsPuzzle[]; extra: GroupsPuzzle[]; holiday?: Record<string, GroupsPuzzle[]> }

export function groupsPuzzleForDay(bank: GroupsBank, day: string, holidays?: HolidayTable | null): GroupsPuzzle | null {
  const pick = bankHolidayPick(day, holidays, bank.holiday);
  if (pick) return pick.entry;
  if (!bank.daily.length) return null;
  return bank.daily[bankIndexForDay(day, bank.daily.length, bank.epoch)] ?? null;
}
export function groupsPuzzleForSeed(bank: GroupsBank, seed: string): GroupsPuzzle | null {
  const pool = bank.extra.length ? bank.extra : bank.daily;
  if (!pool.length) return null;
  return pool[bankIndexForSeed(seed, pool.length)] ?? null;
}
export function groupsDailyNumber(day: string): number {
  const idx = bankDayIndex(day, GROUPS_DAILY_EPOCH);
  return idx === null ? 1 : Math.max(1, idx + 1);
}

/** Fisher–Yates from the end with j = rng() mod (i + 1) — the same shuffle on every platform. */
export function groupsShuffle<T>(items: readonly T[], rng: () => number): T[] {
  const a = items.slice();
  for (let i = a.length - 1; i > 0; i--) {
    const j = rng() % (i + 1);
    [a[i], a[j]] = [a[j], a[i]];
  }
  return a;
}
/** The sixteen words in puzzle order (tier 1 first), then dealt by the seed. */
export function groupsTileOrder(p: GroupsPuzzle, seed: string): string[] {
  const words = [...p.groups].sort((a, b) => a.tier - b.tier).flatMap((g) => g.words);
  return groupsShuffle(words, mulberry32(simpleHash(`${seed}-groups-v1`)));
}

// ── Reducer ────────────────────────────────────────────────────────────────

export type GroupsStatus = 'playing' | 'won' | 'lost';
export type GroupsResult = 'correct' | 'oneaway' | 'wrong' | 'repeat' | 'short';

export interface GroupsState {
  seed: string;
  id: string;
  groups: GroupsGroup[];
  /** Unsolved words in display order. */
  tiles: string[];
  /** Groups in the order solved. */
  solved: GroupsGroup[];
  selected: string[];
  mistakes: number;
  submissions: number;
  hintsUsed: number;
  /** Tiers whose label has been revealed by a hint. */
  revealedTiers: number[];
  /** Pairs shown by hints. */
  pairs: string[][];
  /** Sets already submitted and wrong ("A|B|C|D" sorted). */
  wrongSets: string[];
  shuffles: number;
  lastResult: GroupsResult | null;
  events: string[];
  status: GroupsStatus;
  ended: boolean;
  startTime: number;
  endTime: number | null;
}

export type GroupsAction =
  | { type: 'TOGGLE'; word: string }
  | { type: 'DESELECT' }
  | { type: 'SHUFFLE' }
  | { type: 'SUBMIT' }
  | { type: 'HINT_LABEL' }
  | { type: 'HINT_PAIR' }
  | { type: 'FINISH' };

export function createGroupsState(p: GroupsPuzzle, seed: string, startTime: number): GroupsState {
  const groups = [...p.groups].sort((a, b) => a.tier - b.tier).map((g) => ({ tier: g.tier, label: g.label, words: g.words.map((w) => w.toUpperCase()) }));
  return {
    seed, id: p.id, groups, tiles: groupsTileOrder({ ...p, groups }, seed), solved: [], selected: [], mistakes: 0, submissions: 0, hintsUsed: 0,
    revealedTiers: [], pairs: [], wrongSets: [], shuffles: 0, lastResult: null, events: [], status: 'playing', ended: false, startTime, endTime: null,
  };
}

const setKey = (words: readonly string[]) => [...words].sort().join('|');
export function groupsUnsolved(s: GroupsState): GroupsGroup[] { return s.groups.filter((g) => !s.solved.some((x) => x.tier === g.tier)); }
/** guess_count for the result row: submissions on a win, groups found + 4 on a loss (in play, the current count). */
export function groupsGuessCount(s: GroupsState): number {
  if (s.status === 'lost') return s.solved.length + GROUPS_MAX_MISTAKES;
  return Math.max(s.submissions, s.solved.length);
}
export function groupsBoardsSolved(s: GroupsState): number { return s.solved.length; }
/** The pair a Show-a-pair hint would ring: the two alphabetically first words of the easiest unsolved group not yet paired. */
export function groupsPairTarget(s: GroupsState): { tier: number; pair: string[] } | null {
  for (const g of groupsUnsolved(s)) {
    const words = [...g.words].sort();
    if (s.pairs.some((p) => words.includes(p[0]) && words.includes(p[1]))) continue;
    return { tier: g.tier, pair: [words[0], words[1]] };
  }
  return null;
}
export function groupsLabelTarget(s: GroupsState): GroupsGroup | null {
  return groupsUnsolved(s).find((g) => !s.revealedTiers.includes(g.tier)) ?? null;
}

export function groupsReduce(s: GroupsState, a: GroupsAction, now = 0): GroupsState {
  if (a.type === 'FINISH') return s.status === 'playing' ? s : { ...s, endTime: s.endTime ?? now };
  if (s.ended) return s;

  switch (a.type) {
    case 'TOGGLE': {
      const w = a.word.toUpperCase();
      if (!s.tiles.includes(w)) return s;
      if (s.selected.includes(w)) return { ...s, selected: s.selected.filter((x) => x !== w), lastResult: null };
      if (s.selected.length >= 4) return s;
      return { ...s, selected: [...s.selected, w], lastResult: null };
    }
    case 'DESELECT':
      return s.selected.length ? { ...s, selected: [], lastResult: null } : s;
    case 'SHUFFLE': {
      const n = s.shuffles + 1;
      const tiles = groupsShuffle(s.tiles, mulberry32(simpleHash(`${s.seed}-groups-shuffle-${n}`)));
      return { ...s, tiles, shuffles: n, events: [...s.events, `~${n}`] };
    }
    case 'SUBMIT': {
      if (s.selected.length !== 4) return { ...s, lastResult: 'short' };
      const key = setKey(s.selected);
      if (s.wrongSets.includes(key)) return { ...s, lastResult: 'repeat', events: [...s.events, `=${key.replace(/\|/g, ',')}`] };
      const unsolved = groupsUnsolved(s);
      const hit = unsolved.find((g) => s.selected.every((w) => g.words.includes(w)));
      const submissions = s.submissions + 1;
      if (hit) {
        const solved = [...s.solved, hit];
        const tiles = s.tiles.filter((w) => !hit.words.includes(w));
        const won = solved.length === s.groups.length;
        return {
          ...s, solved, tiles, selected: [], submissions, lastResult: 'correct',
          events: [...s.events, `+${hit.tier}:${hit.words.join(',')}`],
          status: won ? 'won' : s.status, ended: won, endTime: won ? now : s.endTime,
        };
      }
      const best = Math.max(...unsolved.map((g) => s.selected.filter((w) => g.words.includes(w)).length));
      const oneAway = best === 3;
      const mistakes = s.mistakes + 1;
      const lost = mistakes >= GROUPS_MAX_MISTAKES;
      return {
        ...s, mistakes, submissions, wrongSets: [...s.wrongSets, key], lastResult: oneAway ? 'oneaway' : 'wrong',
        selected: lost ? [] : s.selected,
        events: [...s.events, `${oneAway ? 'x1' : 'x0'}:${[...s.selected].sort().join(',')}`],
        status: lost ? 'lost' : s.status, ended: lost, endTime: lost ? now : s.endTime,
      };
    }
    case 'HINT_LABEL': {
      const g = groupsLabelTarget(s);
      if (!g) return s;
      return { ...s, revealedTiers: [...s.revealedTiers, g.tier], hintsUsed: s.hintsUsed + 1, lastResult: null, events: [...s.events, `?c${g.tier}`] };
    }
    case 'HINT_PAIR': {
      const t = groupsPairTarget(s);
      if (!t) return s;
      return { ...s, pairs: [...s.pairs, t.pair], hintsUsed: s.hintsUsed + 2, lastResult: null, events: [...s.events, `?p:${t.pair.join(',')}`] };
    }
    default:
      return s;
  }
}

// ── Matches row ↔ state ────────────────────────────────────────────────────

/** solutions = ["tier|LABEL|W1,W2,W3,W4" × 4, tier order]; guesses = the event log. */
export function groupsMatchRow(s: GroupsState): { solutions: string[]; guesses: string[] } {
  return { solutions: s.groups.map((g) => `${g.tier}|${g.label}|${g.words.join(',')}`), guesses: [...s.events] };
}

export interface GroupsReconstruction {
  groups: GroupsGroup[];
  /** Tiers in the order solved. */
  solvedTiers: number[];
  mistakes: number;
  oneAways: number;
  submissions: number;
  hintsUsed: number;
  revealedTiers: number[];
  pairs: string[][];
  solved: boolean;
}

export function reconstructGroups(solutions: string[] | null | undefined, guesses: string[] | null | undefined): GroupsReconstruction | null {
  if (!solutions || solutions.length !== 4) return null;
  const groups: GroupsGroup[] = [];
  for (const sol of solutions) {
    const parts = sol.split('|');
    if (parts.length !== 3) return null;
    const tier = Number(parts[0]), words = parts[2].split(',');
    if (!(tier >= 1 && tier <= 4) || words.length !== 4) return null;
    groups.push({ tier, label: parts[1], words });
  }
  const solvedTiers: number[] = [], revealedTiers: number[] = [], pairs: string[][] = [];
  let mistakes = 0, oneAways = 0, submissions = 0, hintsUsed = 0;
  for (const ev of guesses ?? []) {
    if (ev.startsWith('+')) { const t = Number(ev.slice(1, ev.indexOf(':'))); if (t >= 1 && t <= 4 && !solvedTiers.includes(t)) solvedTiers.push(t); submissions++; }
    else if (ev.startsWith('x1')) { mistakes++; oneAways++; submissions++; }
    else if (ev.startsWith('x0')) { mistakes++; submissions++; }
    else if (ev.startsWith('?c')) { const t = Number(ev.slice(2)); if (t >= 1 && t <= 4) revealedTiers.push(t); hintsUsed += 1; }
    else if (ev.startsWith('?p:')) { pairs.push(ev.slice(3).split(',')); hintsUsed += 2; }
  }
  return { groups, solvedTiers, mistakes, oneAways, submissions, hintsUsed, revealedTiers, pairs, solved: solvedTiers.length === 4 };
}

File: apps/android/core/src/main/kotlin/com/wordocious/core/Groups.kt
package com.wordocious.core

import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json

/**
 * Kindred — groups of four (More Games §14). 1:1 port of
 * packages/core/src/games/groups.ts; pinned by groups-fixtures.json
 * (GroupsFixtureTest). Sixteen words hide four groups of four; find them all
 * with at most four mistakes. Tiers 1–4 run from a plain category to wordplay
 * and are shown as one to four pips on the solved bar (never color alone).
 * Submitting four words: a group → it locks and its bar appears; three from one
 * group → "One away"; otherwise a plain miss. A set already tried is free to
 * try again (no second mistake). Four mistakes lose the puzzle and the
 * remaining groups are revealed.
 *
 * Hints: NAME A CATEGORY (1 hint) shows the label of the easiest unsolved
 * group; SHOW A PAIR (2 hints) rings two words that belong together. Neither
 * costs a mistake.
 *
 * Result row: guess_count = submissions on a win (4 perfect … 7 worst) and
 * groups found + 4 on a loss; boards_solved = groups found of 4. Event sigils
 * (§11): "+t:W1,W2,W3,W4" solved tier t, "x1:…" one away, "x0:…" miss,
 * "=…" a repeated set (free), "?ct" category hint, "?p:A,B" pair hint,
 * "~n" shuffle n. Tile order comes from mulberry32(simpleHash(seed + "-groups-v1"))
 * so every platform deals the same board. On a holiday (§20) the daily comes
 * from the bank's own holiday list.
 */
const val GROUPS_DAILY_EPOCH = "2026-09-23"
const val GROUPS_MAX_MISTAKES = 4
const val GROUPS_TOTAL_BOARDS = 4
const val GROUPS_PERFECT_GUESSES = 4

@Serializable
data class GroupsGroup(val tier: Int, val label: String, val words: List<String>)

@Serializable
data class GroupsPuzzle(val id: String, val groups: List<GroupsGroup>, val holiday: String? = null)

@Serializable
data class GroupsBank(
    val version: Int, val epoch: String, val daily: List<GroupsPuzzle>, val extra: List<GroupsPuzzle>,
    val holiday: Map<String, List<GroupsPuzzle>>? = null,
) {
    companion object {
        private val json = Json { ignoreUnknownKeys = true }
        fun parse(text: String): GroupsBank? = runCatching { json.decodeFromString<GroupsBank>(text) }.getOrNull()
        /** The bundled bank (core resources/data/groups-puzzles.json, sha-guarded to match the web copy). */
        val bundled: GroupsBank? by lazy {
            GroupsBank::class.java.classLoader?.getResourceAsStream("data/groups-puzzles.json")?.bufferedReader()?.use { it.readText() }?.let { parse(it) }
        }
    }
}

/** The daily puzzle for [day]: the holiday's own entry when the calendar names one, else epoch-indexed. */
fun groupsPuzzleForDay(bank: GroupsBank, day: String, holidays: HolidayTable? = null): GroupsPuzzle? {
    val pick = bankHolidayPick(day, holidays, bank.holiday)
    if (pick != null) return pick.entry
    if (bank.daily.isEmpty()) return null
    return bank.daily[Bank.indexForDay(day, bank.daily.size, bank.epoch)]
}
fun groupsPuzzleForSeed(bank: GroupsBank, seed: String): GroupsPuzzle? {
    val pool = if (bank.extra.isEmpty()) bank.daily else bank.extra
    return if (pool.isEmpty()) null else pool[Bank.indexForSeed(seed, pool.size)]
}
fun groupsDailyNumber(day: String): Int { val idx = Bank.dayIndex(day, GROUPS_DAILY_EPOCH) ?: return 1; return maxOf(1, idx + 1) }

/**
 * Fisher–Yates from the end with j = rng() mod (i + 1) — the same shuffle on
 * every platform. [rng] yields the PRNG's unsigned 32-bit output as a Long
 * (what [Mulberry32.nextU32] returns), so the modulo never sees a negative Int.
 */
fun <T> groupsShuffle(items: List<T>, rng: () -> Long): List<T> {
    val a = ArrayList(items)
    for (i in a.size - 1 downTo 1) {
        val j = (rng() % (i + 1)).toInt()
        val tmp = a[i]; a[i] = a[j]; a[j] = tmp
    }
    return a
}
private fun groupsRng(key: String): () -> Long { val rng = Mulberry32(simpleHash(key)); return { rng.nextU32() } }

/** The sixteen words in puzzle order (tier 1 first), then dealt by the seed. */
fun groupsTileOrder(p: GroupsPuzzle, seed: String): List<String> {
    val words = p.groups.sortedBy { it.tier }.flatMap { it.words }
    return groupsShuffle(words, groupsRng("$seed-groups-v1"))
}

// ── Reducer ──────────────────────────────────────────────────────────────

enum class GroupsStatus(val key: String) { PLAYING("playing"), WON("won"), LOST("lost") }
enum class GroupsResult(val key: String) { CORRECT("correct"), ONEAWAY("oneaway"), WRONG("wrong"), REPEAT("repeat"), SHORT("short") }

/** The pair a Show-a-pair hint would ring, and the tier it belongs to. */
data class GroupsPairTarget(val tier: Int, val pair: List<String>)

/**
 * [tiles]: unsolved words in display order. [solved]: groups in the order solved.
 * [revealedTiers]: tiers whose label has been revealed by a hint. [pairs]: pairs
 * shown by hints. [wrongSets]: sets already submitted and wrong ("A|B|C|D" sorted).
 */
data class GroupsState(
    val seed: String, val id: String, val groups: List<GroupsGroup>, val tiles: List<String>, val solved: List<GroupsGroup>,
    val selected: List<String>, val mistakes: Int, val submissions: Int, val hintsUsed: Int, val revealedTiers: List<Int>,
    val pairs: List<List<String>>, val wrongSets: List<String>, val shuffles: Int, val lastResult: GroupsResult?, val events: List<String>,
    val status: GroupsStatus, val ended: Boolean, val startTime: Long, val endTime: Long?,
) {
    val unsolved: List<GroupsGroup> get() = groupsUnsolved(this)
    val guessCount: Int get() = groupsGuessCount(this)
    val boardsSolved: Int get() = groupsBoardsSolved(this)
    val pairTarget: GroupsPairTarget? get() = groupsPairTarget(this)
    val labelTarget: GroupsGroup? get() = groupsLabelTarget(this)
}

sealed class GroupsAction {
    data class Toggle(val word: String) : GroupsAction()
    object Deselect : GroupsAction()
    object Shuffle : GroupsAction()
    object Submit : GroupsAction()
    object HintLabel : GroupsAction()
    object HintPair : GroupsAction()
    object Finish : GroupsAction()
}

fun createGroupsState(p: GroupsPuzzle, seed: String, startTime: Long): GroupsState {
    val groups = p.groups.sortedBy { it.tier }.map { GroupsGroup(it.tier, it.label, it.words.map { w -> w.uppercase() }) }
    return GroupsState(
        seed, p.id, groups, groupsTileOrder(GroupsPuzzle(p.id, groups, p.holiday), seed), emptyList(), emptyList(), 0, 0, 0,
        emptyList(), emptyList(), emptyList(), 0, null, emptyList(), GroupsStatus.PLAYING, false, startTime, null,
    )
}

private fun setKey(words: List<String>): String = words.sorted().joinToString("|")
fun groupsUnsolved(s: GroupsState): List<GroupsGroup> = s.groups.filter { g -> s.solved.none { it.tier == g.tier } }
/** guess_count for the result row: submissions on a win, groups found + 4 on a loss (in play, the current count). */
fun groupsGuessCount(s: GroupsState): Int {
    if (s.status == GroupsStatus.LOST) return s.solved.size + GROUPS_MAX_MISTAKES
    return maxOf(s.submissions, s.solved.size)
}
fun groupsBoardsSolved(s: GroupsState): Int = s.solved.size
/** The pair a Show-a-pair hint would ring: the two alphabetically first words of the easiest unsolved group not yet paired. */
fun groupsPairTarget(s: GroupsState): GroupsPairTarget? {
    for (g in groupsUnsolved(s)) {
        val words = g.words.sorted()
        if (s.pairs.any { p -> p[0] in words && p[1] in words }) continue
        return GroupsPairTarget(g.tier, listOf(words[0], words[1]))
    }
    return null
}
fun groupsLabelTarget(s: GroupsState): GroupsGroup? = groupsUnsolved(s).firstOrNull { it.tier !in s.revealedTiers }

fun groupsReduce(s: GroupsState, a: GroupsAction, now: Long = 0): GroupsState {
    if (a is GroupsAction.Finish) return if (s.status == GroupsStatus.PLAYING) s else s.copy(endTime = s.endTime ?: now)
    if (s.ended) return s
    return when (a) {
        is GroupsAction.Toggle -> {
            val w = a.word.uppercase()
            if (w !in s.tiles) return s
            if (w in s.selected) return s.copy(selected = s.selected.filter { it != w }, lastResult = null)
            if (s.selected.size >= 4) return s
            s.copy(selected = s.selected + w, lastResult = null)
        }
        is GroupsAction.Deselect -> if (s.selected.isNotEmpty()) s.copy(selected = emptyList(), lastResult = null) else s
        is GroupsAction.Shuffle -> {
            val n = s.shuffles + 1
            val tiles = groupsShuffle(s.tiles, groupsRng("${s.seed}-groups-shuffle-$n"))
            s.copy(tiles = tiles, shuffles = n, events = s.events + "~$n")
        }
        is GroupsAction.Submit -> {
            if (s.selected.size != 4) return s.copy(lastResult = GroupsResult.SHORT)
            val key = setKey(s.selected)
            if (key in s.wrongSets) return s.copy(lastResult = GroupsResult.REPEAT, events = s.events + "=${key.replace("|", ",")}")
            val unsolved = groupsUnsolved(s)
            val hit = unsolved.firstOrNull { g -> s.selected.all { it in g.words } }
            val submissions = s.submissions + 1
            if (hit != null) {
                val solved = s.solved + hit
                val tiles = s.tiles.filter { it !in hit.words }
                val won = solved.size == s.groups.size
                return s.copy(
                    solved = solved, tiles = tiles, selected = emptyList(), submissions = submissions, lastResult = GroupsResult.CORRECT,
                    events = s.events + "+${hit.tier}:${hit.words.joinToString(",")}",
                    status = if (won) GroupsStatus.WON else s.status, ended = won, endTime = if (won) now else s.endTime,
                )
            }
            val best = unsolved.maxOfOrNull { g -> s.selected.count { it in g.words } } ?: -1
            val oneAway = best == 3
            val mistakes = s.mistakes + 1
            val lost = mistakes >= GROUPS_MAX_MISTAKES
            s.copy(
                mistakes = mistakes, submissions = submissions, wrongSets = s.wrongSets + key,
                lastResult = if (oneAway) GroupsResult.ONEAWAY else GroupsResult.WRONG,
                selected = if (lost) emptyList() else s.selected,
                events = s.events + "${if (oneAway) "x1" else "x0"}:${s.selected.sorted().joinToString(",")}",
                status = if (lost) GroupsStatus.LOST else s.status, ended = lost, endTime = if (lost) now else s.endTime,
            )
        }
        is GroupsAction.HintLabel -> {
            val g = groupsLabelTarget(s) ?: return s
            s.copy(revealedTiers = s.revealedTiers + g.tier, hintsUsed = s.hintsUsed + 1, lastResult = null, events = s.events + "?c${g.tier}")
        }
        is GroupsAction.HintPair -> {
            val t = groupsPairTarget(s) ?: return s
            s.copy(pairs = s.pairs + listOf(t.pair), hintsUsed = s.hintsUsed + 2, lastResult = null, events = s.events + "?p:${t.pair.joinToString(",")}")
        }
        is GroupsAction.Finish -> s
    }
}

// ── Matches row ↔ state ─────────────────────────────────────────────────

/** solutions = ["tier|LABEL|W1,W2,W3,W4" × 4, tier order]; guesses = the event log. */
fun groupsMatchRow(s: GroupsState): Pair<List<String>, List<String>> =
    s.groups.map { "${it.tier}|${it.label}|${it.words.joinToString(",")}" } to s.events.toList()

/** [solvedTiers]: tiers in the order solved. */
data class GroupsReconstruction(
    val groups: List<GroupsGroup>, val solvedTiers: List<Int>, val mistakes: Int, val oneAways: Int, val submissions: Int,
    val hintsUsed: Int, val revealedTiers: List<Int>, val pairs: List<List<String>>, val solved: Boolean,
)

fun reconstructGroups(solutions: List<String>?, guesses: List<String>?): GroupsReconstruction? {
    if (solutions == null || solutions.size != 4) return null
    val groups = ArrayList<GroupsGroup>()
    for (sol in solutions) {
        val parts = sol.split("|")
        if (parts.size != 3) return null
        val tier = parts[0].trim().toIntOrNull() ?: return null
        val words = parts[2].split(",")
        if (tier !in 1..4 || words.size != 4) return null
        groups.add(GroupsGroup(tier, parts[1], words))
    }
    val solvedTiers = ArrayList<Int>(); val revealedTiers = ArrayList<Int>(); val pairs = ArrayList<List<String>>()
    var mistakes = 0; var oneAways = 0; var submissions = 0; var hintsUsed = 0
    for (ev in guesses ?: emptyList()) {
        if (ev.startsWith("+")) {
            // TS: Number(ev.slice(1, ev.indexOf(':'))) — a missing ':' slices to -1 (drops the last char).
            val colon = ev.indexOf(':')
            val end = if (colon < 0) maxOf(1, ev.length - 1) else colon
            val t = ev.substring(1, end).trim().toIntOrNull()
            if (t != null && t in 1..4 && t !in solvedTiers) solvedTiers.add(t)
            submissions++
        } else if (ev.startsWith("x1")) { mistakes++; oneAways++; submissions++ }
        else if (ev.startsWith("x0")) { mistakes++; submissions++ }
        else if (ev.startsWith("?c")) { val t = ev.substring(2).trim().toIntOrNull(); if (t != null && t in 1..4) revealedTiers.add(t); hintsUsed += 1 }
        else if (ev.startsWith("?p:")) { pairs.add(ev.substring(3).split(",")); hintsUsed += 2 }
    }
    return GroupsReconstruction(groups, solvedTiers, mistakes, oneAways, submissions, hintsUsed, revealedTiers, pairs, solvedTiers.size == 4)
}
